File: kotlin/src/P04PartialApplication.kt
import kotlin.test.assertEquals

object P04PartialApplication {
    @JvmStatic
    fun main(args: Array<String>) {
        val partial1 = f(1)
        println(partial1.javaClass)
        println(partial1)

        val partial1_3 = partial1(3)
        assertEquals(4, partial1_3)

        assertEquals(9, f(4)(5))
    }

    val f: (Int) -> ((Int) -> Int) = { x -> { y -> x + y } }
    fun g(x: Int, y: Int) = x + y
}

File: haskell/P04PartialApplication.hs
import Data.Typeable
import Test.HUnit.Base

main :: IO ()
main = do
  show (typeOf partial1) @?= "Int -> Int"
  partial1_3 @?= 4
  f 4 5 @?= 9

partial1 = f 1

partial1_3 = partial1 3

f :: Int -> Int -> Int
f x y = x + y

g x y = x + y